### ISSUE TITLE
Out-of-box integration with istio

### ISSUE THREAD

**Issue by mumoshu** [kubernetes-incubator/kube-aws#978]
As mentioned in our [roadmap](https://github.com/kubernetes-incubator/kube-aws/blob/master/ROADMAP.md), let's address this after we bump the default Kubernetes version to v1.8, which supports the Initializers admission control required by automatic-injection of istio sidercars.

**Comment by mumoshu** [comment]
I began to think that we should wait at least until istio v0.6+ for:

- [Autoscaling in Istio using horizontal pod autoscaler](https://github.com/istio/istio/issues/1914)
- [Support for Istio upgrade in Kubernetes](https://github.com/istio/istio/issues/475)
- [Istio Scale & Performance · Issue #369 · istio/istio](https://github.com/istio/istio/issues/369)
- [Lots of broken pipe messages · Issue #2728 · istio/istio](https://github.com/istio/istio/issues/2728)
- [Scale issues in istio-pilot discovery service, affecting alpha perf targets and installation · Issue #1485 · istio/istio](https://github.com/istio/istio/issues/1485#issuecomment-357033179)

Wish I could have more time to actively contribute to Istio but at this point I don't have enough confidence to recommend it for kube-aws users (as, literally, we produce kube-aws for production usage).

**Comment by cmcconnell1** [comment]
Hello All, I was looking into getting this working with the most recent kube-aws release:
`kube-aws version v0.10.2`
Sadly, the [Istio docs only mention requisite configurations for Kops in AWS](https://preliminary.istio.io/docs/setup/kubernetes/platform-setup/aws/)

I asked the Istio Users and Developers groups for any info and also posted in
[istio rocket chat forum](https://istio.rocket.chat/channel/general)
Specifically if anyone in the groups or chat channel had information on how to integrate istio with default vanilla kubernetes (if not kube-aws), etc.
Tl;DR: Only got one response (which just stated "you need to get the order correct") from all of those forums. Not much to go on for me.

I also pinged the  [kube-aws slack channel re istio/kube-aws integration thread:](https://kubernetes.slack.com/archives/C5GP8LPEC/p1537230736000100)
I got some great feedback and thanks again @c-knowles
Must add the below to cluster.yaml and then `kube-aws update`
```
    experimental:
    mutatingAdmissionWebhook:
      enabled: true
    validatingAdmissionWebhook:
      enabled: true
```
Okay, so moving along, getting back to istio docs...
[ref istio docs side-car-injecttion](https://preliminary.istio.io/docs/setup/kubernetes/sidecar-injection/#automatic-sidecar-injection)

> Sidecars can be automatically added to applicable Kubernetes pods using a mutating webhook admission controller. This feature requires Kubernetes 1.9 or later. Verify that the kube-apiserver process has the admission-control flag set with the MutatingAdmissionWebhook and ValidatingAdmissionWebhook admission controllers added and listed in the correct order and the admissionregistration API is enabled.
```
# istio docs
$ kubectl api-versions | grep admissionregistration
admissionregistration.k8s.io/v1alpha1
admissionregistration.k8s.io/v1beta1
```
>

So, now that I've added the initial requisite cluster.yaml file tweaks from Chris in the above kube-aws thread, and comparing against the above Istio doc specified output, I only have
```
# my kube-aws cluster
kubectl api-versions | grep admissionregistration
admissionregistration.k8s.io/v1beta1
```

'Not knowing how to hack the next step, I looked at the Istio AWS Kops doc (link above) and thought I might be able to proceed by taking the Istio Kops-specific instructions and trying to do the equivalent on my kube-aws provisioined cluster, but could use some advice, confirmation,  etc.  I.e.: 

Below [istio aws platform setup only shows kops](https://preliminary.istio.io/docs/setup/kubernetes/platform-setup/aws/) 
It does have a different order for the kubeAPIServer than my kube-aws provisioned cluster, 
What is the process to safely hack/modify this?
> 
> (For Kops) Add the following in the configuration file:
> ```
> kubeAPIServer:
>     admissionControl:
>     - NamespaceLifecycle
>     - LimitRanger
>     - ServiceAccount
>     - PersistentVolumeLabel
>     - DefaultStorageClass
>     - DefaultTolerationSeconds
>     - MutatingAdmissionWebhook
>     - ValidatingAdmissionWebhook
>     - ResourceQuota
>     - NodeRestriction
>     - Priority
> ```
> 

Looking at my kube-aws provisioned cluster, I do see a different order with the kube-apiserver
```
for i in $(kk get pods | grep 'kube-apiserver' | awk '{print $1}'); do kk describe pods $i | grep -e "(Node|admission-control)"; done
Node:         ip-10-1-8-102.us-west-1.compute.internal/10.1.8.102
      --authorization-mode=Node,RBAC
      --admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,NodeRestriction,ResourceQuota,DefaultTolerationSeconds,MutatingAdmissionWebhook,ValidatingAdmissionWebhook
Node-Selectors:    <none>
Node:         ip-10-1-8-68.us-west-1.compute.internal/10.1.8.68
      --authorization-mode=Node,RBAC
      --admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,NodeRestriction,ResourceQuota,DefaultTolerationSeconds,MutatingAdmissionWebhook,ValidatingAdmissionWebhook
Node-Selectors:    <none>
```

If I could get help getting Istio && kube-aws integration working and get your confirmation(s) that the proposed/recommended process is the "kube-aws approved method", I could then do a P/R for the doc enhancement update to support that functionality.

Thank you

**Comment by c-knowles** [comment]
@cmcconnell1 any particular problem you are facing? I'm not sure if the alpha admission registration is needed for Istio still?

If you do need that, I dug a little bit and found this:
https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#enable-initializers-alpha-feature

That is reflected in this line:
https://github.com/kubernetes-incubator/kube-aws/blob/e8d6bea5760fb07a0b5cf83399fc6f0ba23fbbd7/core/controlplane/config/templates/cloud-config-controller#L3073

So can be enabled using this in `cluster.yaml`:
```
experimental:
  admission:
    initializers:
      enabled: true
```

**Comment by c-knowles** [comment]
I believe Istio uses to use Initializers but swapped to webhook admission which is why I'm not sure if it's needed any more.

I agree better out of the box integration using the kube-aws plugin support would be good.

**Comment by cmcconnell1** [comment]
Thanks for your follow-up.  I made those changes to the config and applied them but haven't got back to istio testing yet.  I've been working on a few POC's for service mesh integration (of course using kube-aws clusters ;-)  ) and wanted to spend some time with the most popular options.  Being new to service mesh in general, I wanted to make sure that things worked as specified before I invested a lot of time and then introduced to our team.  As often the case, the docs might be a bit out of date on the Istio side.   

However, what I can say for me is that I really like Linkerd having a three word cluster validation command
[Step 2: Validate your Kubernetes cluster](https://linkerd.io/2/getting-started/#step-2-validate-your-kubernetes-cluster)

> Kubernetes clusters can be configured in many different ways. To ensure that the control plane will install correctly, the Linkerd CLI can check and validate that everything is configured correctly.
> To check that your cluster is configured correctly and ready to install the control plane, you can run:
> `linkerd check --pre`

I would appreciate recommendations and opinions on both Istio and Linkerd service mesh options.  I.e. was there a reason that Istio was targeted for kube-aws integration instead of Linkerd, etc?  This would be very helpful for us trying to [herd the cats](https://en.wikipedia.org/wiki/Herding_cats) of kubernetes ecosystem components, integrations, etc. 

Thank you

**Comment by mumoshu** [comment]
@cmcconnell1 @c-knowles Thanks a lot for chiming in, and your supports!

> was there a reason that Istio was targeted for kube-aws integration instead of Linkerd, etc?

This is my personal opinion. I thought it makes sense to provide a smooth path for integrating kube-aws and istio, rather than linkerd, because istio seems to cover much many use-cases and linkerd is already super easy to get started.

Regarding the covered use-cases, I think istio's benefits over linkered(2) are:

- cross-cluster mesh
- automatic mutual tls
- cilium integration
- serving as a platform for other high-level platforms like knative

The cross-cluster mesh sounds especially useful FOR ME because then I can use istio as the global load balancer that can be relatively safely rolling-updated.

The global LB can then be used as a solid foundation to achieve blue-green deployments of kube-aws clusters with smooth traffic shifting. It will drastically improve how we could continuously and automatically upgrade kube-aws clusters #1050.

**Comment by davidmccormick** [comment]
Good morning!  We are trialing Istio in a few of our kube-aws clusters and to the best of my knowledge all we had to do is enable the mutating and validating webhooks and installed istio via helm charts to get it up and running.

I do think that both out-of-the-box Istio and ingress controllers would be great additions to kube-aws and allow users to bootstrap fuller, more useful clusters more easily.  I appreciate that these are areas where there are admittedly quite a few options and lots of movement but I think it would be good for us to offer a couple of opinionated choices - much like we do with networking.
